qcengine/programs/openmm.py

```diff
@@ -0,0 +1,194 @@
+
+        # Build out the base molecule
+        base_mol = Chem.Mol()
+        rw_mol = Chem.RWMol(base_mol)
+        for sym in jmol.symbols:
+            rw_mol.AddAtom(Chem.Atom(sym.title()))
+
+        # Add in connectivity
+        bond_types = {1: Chem.BondType.SINGLE, 2: Chem.BondType.DOUBLE, 3: Chem.BondType.TRIPLE}
+        for atom1, atom2, bo in jmol.connectivity:
+            rw_mol.AddBond(atom1, atom2, bond_types[bo])
+
+        mol = rw_mol.GetMol()
+
+        # Write out the conformer
+        natom = len(jmol.symbols)
+        conf = Chem.Conformer(natom)
+        bohr2ang = ureg.conversion_factor("bohr", "angstrom")
+        for line in range(natom):
+            conf.SetAtomPosition(line, (bohr2ang * jmol.geometry[line, 0],
+                                        bohr2ang * jmol.geometry[line, 1],
+                                        bohr2ang * jmol.geometry[line, 2]))  # yapf: disable
+
+        mol.AddConformer(conf)
+        Chem.rdmolops.SanitizeMol(mol)
+
+        return mol
+
+    def compute(self, input_data: 'ResultInput', config: 'JobConfig') -> 'Result':
+        """
+        Runs OpenMM on given structure, inputs, in vacuum.
+        """
+        self.found(raise_error=True)
+
+        from simtk import openmm
+        from simtk import unit
+
+        from openforcefield.topology import Molecule
+        from openforcefield.typing.engines.smirnoff import ForceField
+        
+        # Failure flag
+        ret_data = {"success": False}
+
+        # get number of threads to use from `JobConfig.ncores`; otherwise, try environment variable
+        nthreads = config.ncores
+        if nthreads is None:
+            nthreads = os.environ.get('OPENMM_CPU_THREADS')
+
+        # Set workdir to scratch
+        # Location resolution order config.scratch_dir, /tmp
+        parent = config.scratch_directory
+        with temporary_directory(parent=parent, suffix="_openmm_scratch") as tmpdir:
+
+            # Grab molecule, forcefield
+            jmol = input_data.molecule
+            
+            # TODO: If urls are supported by
+            # `openforcefield.typing.engines.smirnoff.ForceField` already, we
+            # can eliminate the `offxml` and `url` distinction
+            # URL processing can happen there
+            if input_data.model.offxml:
+                offxml = input_data.model.offxml
+            elif input_data.model.url:
+                # TODO: add pull of offxml data from a url
+                pass
+            else:
+                raise InputError("OpenMM requires either `model.offxml` or `model.url` to be set")
+
+            # Process molecule with RDKit
+            rdkit_mol = self._process_molecule_rdkit(jmol)
+
+            # Create an OFF Molecule from the RDKit Molecule
+            off_mol = Molecule(rdkit_mol)
+
+            # Load an OFF Force Field
+            off_forcefield = ForceField(offxml)
```

Comment:
Creating a new `ForceField` object for every configuration of every molecule we want to evaluate will add a huge amount of overhead. Is there a way we can either memoize/cache this creation or somehow refactor things to allow this engine to apply the force field to many molecules once the `off_forcefield` object is created?

Reply:
I'm not aware of where QCEngine implements caching. Is this something that currently exists? If so, can you point me to where it is currently used? I can then use this as an example to start from.

Reply:
I don't think QCEngine naturally implements caching since generally the computations take minutes to hours. Since these OpenMM calculations take milliseconds, it may make sense to try to minimize the overhead here. We can certainly implement that later if needed.

Reply:
See the ANI Harness where we cache the pytorch graph.

Reply:
See [here](https://github.com/MolSSI/QCEngine/blob/master/qcengine/programs/torchani.py#L17-L75).

Reply:
LRU cache implemented for OFF `Forcefield`.